interpreter/attributes.go

```diff
@@ -0,0 +1,186 @@
+// Copyright 2019 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package interpreter
+
+import (
+	"fmt"
+
+	"github.com/google/cel-go/common/types"
+	"github.com/google/cel-go/common/types/ref"
+)
+
+// Attribute refers to a Variable and zero or more field or index traversals.
+type Attribute interface {
+	// Variable referenced by the attribute.
+	//
+	// When the Attribute is absolute, the Variable value is non-nil. When the Attribute is
+	// relative such as might happen when refering to the field or index of a object-based
+	// return value from a function, the Variable value will be nil.
+	Variable() Variable
+
+	// Path of field traversals against reference object, either a variable or a contextual
+	// intermediate value computed during CEL evaluation.
+	//
+	// If Variable is nil, the Path must contain at least one PathElem.
+	Path() []*PathElem
+
+	// Select a path element from the current Attribute to produce a more fully qualified
+	// Attribute.
+	Select(*PathElem) Attribute
+}
+
+// Variable refers to a type-level expression identifier corresponding to program input.
+type Variable interface {
+	// ID indicates the expression id where the variable appears.
+	//
+	// Non-zero when the Variable is identified during program plan time.
+	//
+	// Zero-ish when the Variable is provided as part of a user-provided Attribute specification
+	// for use with partial state evaluation.
+	ID() int64
+
+	// Name of the variable, may be simple or namespaced, e.g. 'in' or 'cloud.iam.resource'.
```

Comment:
`in` is a keyword, so cannot be a variable, right?

Reply:
Good point, I'll update the comment.